experimental/lib/Support/FtdSupport.cpp

```diff
@@ -14,24 +14,91 @@
 
 #include "experimental/Support/FtdSupport.h"
 #include "mlir/Dialect/ControlFlow/IR/ControlFlowOps.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
 #include <unordered_set>
 
 using namespace mlir;
 using namespace dynamatic;
-using namespace experimental::boolean;
+using namespace dynamatic::experimental;
 
-/// Recursive function which allows to obtain all the paths from block "start"
-/// to block "end" using a DFS, possibly traversing "blockToTraverse" and not
-/// traversing "blocksToAvoid".
+std::string ftd::BlockIndexing::getBlockCondition(Block *block) const {
+  return "c" + std::to_string(getIndexFromBlock(block));
+}
+
+ftd::BlockIndexing::BlockIndexing(Region &region) {
+  mlir::DominanceInfo domInfo;
+
+  // Create a vector with all the blocks
+  SmallVector<Block *> allBlocks;
+  for (Block &bb : region.getBlocks())
+    allBlocks.push_back(&bb);
+
+  // Sort the vector according to the dominance information, so that a block
+  // comes before each dominators.
+  std::sort(allBlocks.begin(), allBlocks.end(),
+            [&](Block *a, Block *b) { return domInfo.dominates(a, b); });
+
+  // Associate a smalled index in the map to the blocks at higer levels of the
+  // dominance tree
+  for (auto [blockID, bb] : llvm::enumerate(allBlocks)) {
+    indexToBlock.insert({blockID, bb});
+    blockToIndex.insert({bb, blockID});
+  }
+}
+
+Block *ftd::BlockIndexing::getBlockFromIndex(unsigned index) const {
+  auto it = indexToBlock.find(index);
+  return (it == indexToBlock.end()) ? nullptr : it->getSecond();
+}
+
+Block *
+ftd::BlockIndexing::getBlockFromCondition(const std::string &condition) const {
+  std::string conditionNumber = condition;
+  conditionNumber.erase(0, 1);
+  unsigned index = std::stoi(conditionNumber);
+  return this->getBlockFromIndex(index);
+}
+
+unsigned ftd::BlockIndexing::getIndexFromBlock(Block *bb) const {
+  auto it = blockToIndex.find(bb);
+  return (it == blockToIndex.end()) ? -1 : it->getSecond();
+}
+
+bool dynamatic::experimental::ftd::BlockIndexing::greaterIndex(
+    Block *bb1, Block *bb2) const {
+  return getIndexFromBlock(bb1) > getIndexFromBlock(bb2);
+}
+
+bool ftd::BlockIndexing::lessIndex(Block *bb1, Block *bb2) const {
+  return getIndexFromBlock(bb1) < getIndexFromBlock(bb2);
+}
+
+/// Recursively check weather 2 blocks belong to the same loop, starting
+/// from the inner-most loops.
+static bool isSameLoop(const CFGLoop *loop1, const CFGLoop *loop2) {
+  if (!loop1 || !loop2)
+    return false;
+  return (loop1 == loop2 || isSameLoop(loop1->getParentLoop(), loop2) ||
+          isSameLoop(loop1, loop2->getParentLoop()) ||
+          isSameLoop(loop1->getParentLoop(), loop2->getParentLoop()));
+}
+
+bool ftd::isSameLoopBlocks(Block *source, Block *dest,
+                           const mlir::CFGLoopInfo &li) {
+  return isSameLoop(li.getLoopFor(source), li.getLoopFor(dest));
```

Comment:
inline in header

Reply:
The used function `isSameLoop` is declared as static in the `.cpp` file, thus I cannot move the declaration in the header...